modyn/trainer_server/internal/dataset/key_sources/local_key_source.py

```diff
@@ -32,3 +32,12 @@ def get_num_data_partitions(self) -> int:
 
     def end_of_trigger_cleaning(self) -> None:
         self._trigger_sample_storage.clean_trigger_data(self._pipeline_id, self._trigger_id)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state["_trigger_sample_storage"]  # not pickable
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self._trigger_sample_storage = TriggerSampleStorage(self.offline_dataset_path)
```

Comment:
when do we pickle/unpickle local_key_source? Didn't find it in any place

Reply:
The trainer server uses multiple data loading processes (workers). When they are started, `local_key_source` is pickled as part of the OnlineDataset being pickled. What was a bit weird to me is that it does not always occur (it should from my understanding), but nevertheless after implementing this it never happened again